oak_functions_utils/src/sync.rs

```diff
@@ -15,10 +15,70 @@
 //
 
 use core::{
+    cell::UnsafeCell,
     hint::spin_loop,
+    mem::MaybeUninit,
     sync::atomic::{AtomicBool, Ordering},
 };
 use lock_api::{GuardSend, RawMutex};
+use spinning_top::RawSpinlock;
+
+/// A synchronised implementation of a cell that can be initialized only once.
+///
+/// This achieves the same goal as `std::sync::OnceLock`, but uses a spinlock to be compatible with
+/// `no_std`.
+pub struct OnceCell<T> {
+    initialized: AtomicBool,
+    lock: RawSpinlock,
+    value: UnsafeCell<MaybeUninit<T>>,
+}
+
+impl<T> OnceCell<T> {
+    pub const fn new() -> Self {
+        Self {
+            initialized: AtomicBool::new(false),
+            lock: RawSpinlock::INIT,
+            value: UnsafeCell::new(MaybeUninit::uninit()),
+        }
+    }
+
+    pub fn get(&self) -> Option<&T> {
+        if !self.initialized.load(Ordering::Acquire) {
+            return None;
+        }
+        // Safety: there are no mutable references to the inner value if `initialized` is set to
+        // `true`, and the value has been initialized, so it is safe to create a new shared
+        // reference and assume it is initialized.
+        Some(unsafe { (*self.value.get()).assume_init_ref() })
+    }
+
+    pub fn set(&self, value: T) -> Result<(), T> {
+        let mut result = Ok(());
+        // Do an initial check to see whether the value has been initialized.
+        if !self.initialized.load(Ordering::Acquire) {
+            // Lock to make sure we have exclusive access.
+            self.lock.lock();
```

Comment:
Would tehere be some kind of RAII way of doing this? https://docs.rs/spinning_top/latest/spinning_top/type.SpinlockGuard.html maybe?

Reply:
Yes, I tried to avoid having a Spinlock<()>, but it is probalby cleaner to have that this the explicit unlock.